### ISSUE TITLE
[pdata] Implement Go 1.23 style iterators

### ISSUE THREAD

**Issue by yurishkuro** [open-telemetry/opentelemetry-collector#11982]
**Is your feature request related to a problem? Please describe.**

Iterating through `pdata` structures today is pretty convoluted, e.g.

```go
for i := 0; i < traces.ResourceSpans().Len(); i++ {
	resource := traces.ResourceSpans().At(i)
	for j := 0; j < resource.ScopeSpans().Len(); j++ {
		scope := resource.ScopeSpans().At(j)
		for k := 0; k < scope.Spans().Len(); k++ {
			span := scope.Spans().At(k)
```

**Describe the solution you'd like**
Go 1.23 introduced the ability to use iterators for looping through custom structs. The above code could look like this, if we introduce iterator functions to pdata, e.g. `Seq()` and `Seq2()`:

```go
for resource := range traces.ResourceSpans().Seq() {
	for scope := range resource.ScopeSpans().Seq() {
		for span := range scope.Spans().Seq() {
```

```go
for i, resource := range traces.ResourceSpans().Seq2() {
	for j, scope := range resource.ScopeSpans().Seq2() {
		for k, span := range scope.Spans().Seq2() {
```

**Comment by bogdandrutu** [comment]
@yurishkuro good call, will do when we move to minimum supported version to be 1.23

**Comment by mauri870** [comment]
I think this is perfect use case for iterators, the way this is implemented today is quite verbose to write and almost every single use case requires looping over the entries in some way.

I suggest we go with an `All` method, following the api signature for the already existing [maps](https://github.com/golang/go/issues/61900) and [slices](https://github.com/golang/go/issues/61899) counterparts.

Since this involves changes in [autogenerated code](https://github.com/open-telemetry/opentelemetry-collector/tree/ba4ab80/pdata/internal/cmd/pdatagen/internal/templates), there might be oportunities for reusing existing iterator routines from the standard library, for example `ResourceSpans` uses an slice internally.

**Comment by mauri870** [comment]
I started working on this https://github.com/open-telemetry/opentelemetry-collector/compare/main...mauri870:opentelemetry-collector:pdata-iterator-all. It requires Go 1.23 so we cannot accept it until all the other collector packages are bumped to 1.23.
